src/Nethermind/Nethermind.Logging.NLog/NLogManager.cs

```diff
@@ -17,8 +17,16 @@ namespace Nethermind.Logging.NLog
     public class NLogManager : ILogManager
     {
         private const string DefaultFileTargetName = "file-async_wrapped";
+        private const string DefaultFolder = "logs";
 
         public NLogManager(string logFileName, string logDirectory = null, string logRules = null)
+        {
+            Setup(logFileName, logDirectory, logRules);
+            // Required since 'NLog.config' could change during runtime, we need to re-apply the configuration
+            LogManager.ConfigurationChanged += (sender, args) => Setup(logFileName, logDirectory, logRules);
```

Comment:
**What are the best practices when dealing with this kind of event handler?**

I've read that you could use `LogManager.ConfigurationChanged = null` to remove all handlers, but what if we attach a new one on other parts of the code?

Ideally, we would only remove this handler during `Shutdown`.

Reply:
is that a static event? and do we want to attach a handler to it every time we create a new NLogManager ?

Reply:
Yes, it is:

```C#
public static event EventHandler<LoggingConfigurationChangedEventArgs> ConfigurationChanged;
```

`NLogManager` is created in two scenarios (ignoring tests):

https://github.com/NethermindEth/nethermind/blob/90a18eb5a45558b0dcbe2b7d438bdb847c22bac1/src/Nethermind/Nethermind.Runner/Program.cs#L101

https://github.com/NethermindEth/nethermind/blob/90a18eb5a45558b0dcbe2b7d438bdb847c22bac1/src/Nethermind/Nethermind.Runner/Program.cs#L163

IMO, it should be treated as a singleton (does it even make sense to have two NLog instances?)

Reply:
Why do you need to unregister it at all if you want it to live throughout the whole application lifecycle?
Or like Tomasz said make `NLogManager` or `ILogManager` implement `IDisposable` unhook there and leave it to user to dispose the manager. (Or if you don't trust your users you can use disposable/finalizer pattern)

Reply:
> Why do you need to unregister it at all if you want it to live throughout the whole application lifecycle?

That's the thing: I do not know if we want/need to unregister it. We already have a shutdown method that kind of looks like `Dispose`, but from reading the source of NLog I don't see that it removes the event handler (https://github.com/NLog/NLog/blob/3ff3d5b2ab091e2171e81c410b4c4aa558aee4ba/src/NLog/LogFactory.cs#L1020).

I think that it should not be problematic to do nothing, but I wanted to double check with someone who has worked with NLog and EventHandlers before.

Reply:
If we are shutting down you can kind of ignore things that are not crucial for shutting down. But Also you can make it IDisposable and add it to DisposeStack - a stack of things that will be disposed during shutdown

Reply:
So I guess our options are:
1. Do nothing with the event handler since it lasts for the whole application lifetime.
2. Use `IDisposable` and the dispose pattern to ensure that we remove the event handler.
3. Change the config to use [Configuration-Variables](https://github.com/NLog/NLog/wiki/Var-Layout-Renderer), removing the requirement of running `SetupLogFile` and `SetupLogDirectory` (but what about `SetupLogRules`?).

I think `1.` should be good enough.

Reply:
I would add IDisposable for completeness sake - so 2.- but shouldn't affect the end result,